File: sdks/python/src/trustlayer/guardian.py
"""Client for the TrustLayer ``cynepic-guardian`` policy service.

The guardian is the Rust HTTP sidecar implemented in ``core-rs``. The contract:

    POST /v1/check  {"event": <AgentTraceEvent>, "policy_name": "default"}
    -> 200 {"decision": "PASS"|"FAIL"|"ESCALATE", "rule": "...", "reason": "...", "policy": "..."}

The client is **fail-open by default** so an unavailable guardian cannot
take down the host agent. Override with ``fail_open=False`` when you need
hard denial (e.g. regulated workloads where missing a policy check is
worse than blocking).
"""

from __future__ import annotations

import json
import logging
import os
from types import TracebackType
from typing import Literal, Self, TypedDict

import httpx

from .schema import AgentTraceEvent, PolicyCheckResult

logger = logging.getLogger("trustlayer.guardian")

DEFAULT_GUARDIAN_ENDPOINT = "http://127.0.0.1:8089/v1/check"
API_TOKEN_ENV_VAR = "TRUSTLAYER_API_TOKEN"


class Verdict(TypedDict):
    decision: Literal["PASS", "FAIL", "ESCALATE"]
    rule: str | None
    reason: str | None
    policy: str


class GuardianClient:
    """Synchronous client for the cynepic-guardian HTTP service.

    On transport or HTTP failure, returns a synthetic ``"policy": "fallback"``
    verdict whose ``decision`` is ``PASS`` (fail-open) or ``FAIL``
    (fail-closed) depending on ``fail_open``.

    The bearer token (ADR-007) resolves in this order:

    1. ``api_key`` argument (explicit wins).
    2. ``TRUSTLAYER_API_TOKEN`` environment variable.
    3. None — no Authorization header sent.
    """

    def __init__(
        self,
        endpoint: str = DEFAULT_GUARDIAN_ENDPOINT,
        *,
        policy_name: str | None = None,
        api_key: str | None = None,
        timeout: float = 1.0,
        fail_open: bool = True,
        transport: httpx.BaseTransport | None = None,
    ) -> None:
        self.endpoint = endpoint
        self.policy_name = policy_name
        self.fail_open = fail_open
        resolved = api_key if api_key is not None else os.environ.get(API_TOKEN_ENV_VAR) or None
        self._client = httpx.Client(
            timeout=timeout,
            transport=transport,
            headers=self._build_headers(resolved),
        )

    @staticmethod
    def _build_headers(api_key: str | None) -> dict[str, str]:
        headers = {"Content-Type": "application/json"}
        if api_key:
            headers["Authorization"] = f"Bearer {api_key}"
        return headers

    def check(
        self,
        event: AgentTraceEvent,
        policy_name: str | None = None,
    ) -> Verdict:
        body = json.dumps(
            {
                "event": json.loads(event.model_dump_json()),
                "policy_name": policy_name or self.policy_name,
            }
        )
        try:
            response = self._client.post(self.endpoint, content=body)
            response.raise_for_status()
            return _coerce_verdict(response.json())
        except (httpx.HTTPError, ValueError) as exc:
            logger.warning("trustlayer-guardian check failed: %s", exc)
            return self._fallback_verdict(str(exc))

    def _fallback_verdict(self, detail: str) -> Verdict:
        decision: Literal["PASS", "FAIL", "ESCALATE"] = (
            PolicyCheckResult.PASS.value if self.fail_open else PolicyCheckResult.FAIL.value
        )
        return Verdict(
            decision=decision,
            rule=None,
            reason=f"guardian unavailable: {detail}",
            policy="fallback",
        )

    def close(self) -> None:
        self._client.close()

    def __enter__(self) -> Self:
        return self

    def __exit__(
        self,
        exc_type: type[BaseException] | None,
        exc: BaseException | None,
        tb: TracebackType | None,
    ) -> None:
        self.close()


def _coerce_verdict(data: object) -> Verdict:
    if not isinstance(data, dict):
        raise ValueError(f"unexpected verdict payload type: {type(data).__name__}")
    decision = data.get("decision")
    if decision not in {"PASS", "FAIL", "ESCALATE"}:
        raise ValueError(f"unexpected verdict decision: {decision!r}")
    return Verdict(
        decision=decision,
        rule=data.get("rule"),
        reason=data.get("reason"),
        policy=str(data.get("policy", "unknown")),
    )

File: sdks/go/trustlayer/guardian.go
package trustlayer

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"time"
)

// DefaultGuardianEndpoint is the loopback URL the reference Rust
// sidecar serves /v1/check on. Override via GuardianOptions.Endpoint.
const DefaultGuardianEndpoint = "http://127.0.0.1:8089/v1/check"

// GuardianOptions configures the cynepic-guardian client.
type GuardianOptions struct {
	// Endpoint is the full URL to POST /v1/check on. Empty = default.
	Endpoint string
	// PolicyName picks a named policy on the server. Empty = server default.
	PolicyName string
	// APIKey is the bearer token (ADR-007). Same resolution order as
	// the ingest client.
	APIKey string
	// HTTPClient lets callers inject a *http.Client. nil installs a
	// fresh client with a 1s timeout (matches the Python default —
	// guardian calls are on the hot path).
	HTTPClient *http.Client
	// FailOpen controls the fallback verdict when the guardian is
	// unreachable. true (default) returns PASS so a guardian outage
	// doesn't take down the host agent.
	FailOpen *bool
}

// Verdict is the guardian's reply (spec §5.2).
type Verdict struct {
	Decision Decision `json:"decision"`
	Rule     *string  `json:"rule,omitempty"`
	Reason   *string  `json:"reason,omitempty"`
	Policy   string   `json:"policy"`
}

// GuardianClient is a goroutine-safe guardian client.
type GuardianClient struct {
	endpoint   string
	policyName string
	apiKey     string
	http       *http.Client
	failOpen   bool
}

// NewGuardian constructs a GuardianClient. Like NewClient, no I/O until
// the first Check.
func NewGuardian(opts GuardianOptions) (*GuardianClient, error) {
	endpoint := opts.Endpoint
	if endpoint == "" {
		endpoint = DefaultGuardianEndpoint
	}
	httpClient := opts.HTTPClient
	if httpClient == nil {
		httpClient = &http.Client{Timeout: time.Second}
	}
	failOpen := true
	if opts.FailOpen != nil {
		failOpen = *opts.FailOpen
	}
	return &GuardianClient{
		endpoint:   endpoint,
		policyName: opts.PolicyName,
		apiKey:     resolveAPIToken(opts.APIKey),
		http:       httpClient,
		failOpen:   failOpen,
	}, nil
}

// Endpoint exposes the configured URL (handy for tests).
func (g *GuardianClient) Endpoint() string { return g.endpoint }

// FailOpen reports the configured fallback posture.
func (g *GuardianClient) FailOpen() bool { return g.failOpen }

// Check forwards an event to the guardian and returns the verdict.
// On any transport / parse failure, returns a synthetic
// {Policy: "fallback"} verdict whose Decision depends on FailOpen.
// The error return is always nil — instrumentation must never propagate
// guardian failures as fatal — but the synthetic verdict's Reason
// carries the detail for logging.
func (g *GuardianClient) Check(ctx context.Context, event AgentTraceEvent) (Verdict, error) {
	body, err := json.Marshal(map[string]any{
		"event":       event,
		"policy_name": policyNameJSON(g.policyName),
	})
	if err != nil {
		return g.fallback(fmt.Sprintf("marshal request: %s", err)), nil
	}
	req, err := http.NewRequestWithContext(ctx, http.MethodPost, g.endpoint, bytes.NewReader(body))
	if err != nil {
		return g.fallback(fmt.Sprintf("build request: %s", err)), nil
	}
	req.Header.Set("Content-Type", "application/json")
	if g.apiKey != "" {
		req.Header.Set("Authorization", "Bearer "+g.apiKey)
	}
	resp, err := g.http.Do(req)
	if err != nil {
		return g.fallback(fmt.Sprintf("transport: %s", err)), nil
	}
	defer resp.Body.Close()
	if resp.StatusCode != http.StatusOK {
		return g.fallback(fmt.Sprintf("HTTP %d", resp.StatusCode)), nil
	}
	var v Verdict
	if err := json.NewDecoder(resp.Body).Decode(&v); err != nil {
		return g.fallback(fmt.Sprintf("decode verdict: %s", err)), nil
	}
	if _, ok := validDecisions[v.Decision]; !ok {
		return g.fallback(fmt.Sprintf("unexpected decision %q", v.Decision)), nil
	}
	if v.Policy == "" {
		v.Policy = "unknown"
	}
	return v, nil
}

// CheckWithPolicy overrides the configured policy_name for one call.
func (g *GuardianClient) CheckWithPolicy(ctx context.Context, event AgentTraceEvent, policyName string) (Verdict, error) {
	g2 := *g
	g2.policyName = policyName
	return g2.Check(ctx, event)
}

// policyNameJSON returns an interface that marshals to null when the
// configured policy_name is empty, so the wire request matches the
// shape sent by the Python and TypeScript SDKs.
func policyNameJSON(name string) any {
	if name == "" {
		return nil
	}
	return name
}

func (g *GuardianClient) fallback(detail string) Verdict {
	decision := DecisionPass
	if !g.failOpen {
		decision = DecisionFail
	}
	reason := fmt.Sprintf("guardian unavailable: %s", detail)
	return Verdict{
		Decision: decision,
		Rule:     nil,
		Reason:   &reason,
		Policy:   "fallback",
	}
}

// Close is a no-op today; exists for API symmetry / future pooling.
func (g *GuardianClient) Close() error { return nil }
